File: c/11.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#include <ctype.h>
#include <math.h>
#define defaultInput "../../Inputs/2021_11.txt"
#include "../../Modules/input.h"
#include "../../Modules/dict.h"


char *pointStr(int x, int y) {
    int sizex = ceil(log10(abs(x) + 1)), sizey = ceil(log10(abs(y) + 1));
    char *str = (char*)calloc(sizex + sizey + 2, sizeof(char));
    sprintf(str, "%d,%d", x, y);
    return str;
}


struct vector *strPoint(char *pStr) {
    struct vector *point = createVector(intsize, copyElement);
    char *cpy = (char*)calloc(strlen(pStr), sizeof(char));
    strcpy(cpy, pStr);

    char *p = strtok(cpy, ",");
    while (p) {
        int *n = (int*)calloc(1, sizeof(int));
        *n = atoi(p);
        appendVector(point, n);

        p = strtok(NULL, ",");
    }

    return point;
}


int flashOcto(struct dict *octopi, int fx, int fy) {
    int flashes = 1;

    for (int xoff = -1; xoff <= 1; xoff++) {
        for (int yoff = -1; yoff <= 1; yoff++) {
            if (xoff == 0 && yoff == 0) {
                continue;
            }

            char *nP = pointStr(fx + xoff, fy + yoff);
            if (inDict(octopi, nP)) {
                int *octo = (int*)getDictVal(octopi, nP);
                *octo = *octo + 1;

                if (*octo == 10) {
                    flashes += flashOcto(octopi, fx + xoff, fy + yoff);
                }
            }
        }
    }

    return flashes;
}

int part1(char *fileName) {
    struct vector *input_data = multiLine(fileName);
    struct dict *octopi = createDict(stringsize, intsize, copyElement);

    for (int y = 0; y < input_data->len; y++) {
        char *line = (char*)input_data->arr[y];
        for (int x = 0; x < strlen(line); x++) {
            char *c = (char*)calloc(2, sizeof(char));
            strncpy(c, line + x, 1);
            int *n = (int*)calloc(1, sizeof(int));
            *n = atoi(c);
            addDict(octopi, pointStr(x, y), n);
        }
    }

    struct vector *locs = keys2vector(octopi);
    int flashes = 0;
    for (int steps = 0; steps < 100; steps++) {
        for (int i = 0; i < locs->len; i++) {
            struct vector *posVec = strPoint((char*)locs->arr[i]);
            int *octo = (int*)getDictVal(octopi, locs->arr[i]), x = *(int*)posVec->arr[0], y = *(int*)posVec->arr[1];
            *octo = *octo + 1;

            if (*octo == 10) {
                flashes += flashOcto(octopi, x, y);
            }
        }

        for (int i = 0; i < locs->len; i++) {
            int *octo = (int*)getDictVal(octopi, locs->arr[i]);

            if (*octo >= 10) {
                *octo = 0;
            }
        }
    }

    return flashes;
}

int part2(char *fileName) {
    struct vector *input_data = multiLine(fileName);
    struct dict *octopi = createDict(stringsize, intsize, copyElement);

    for (int y = 0; y < input_data->len; y++) {
        char *line = (char*)input_data->arr[y];
        for (int x = 0; x < strlen(line); x++) {
            char *c = (char*)calloc(2, sizeof(char));
            strncpy(c, line + x, 1);
            int *n = (int*)calloc(1, sizeof(int));
            *n = atoi(c);
            addDict(octopi, pointStr(x, y), n);
        }
    }

    struct vector *locs = keys2vector(octopi);
    int steps = 0, flashes = 0;
    while (true) {
        for (int i = 0; i < locs->len; i++) {
            struct vector *posVec = strPoint((char*)locs->arr[i]);
            int *octo = (int*)getDictVal(octopi, locs->arr[i]), x = *(int*)posVec->arr[0], y = *(int*)posVec->arr[1];
            *octo = *octo + 1;

            if (*octo == 10) {
                flashes += flashOcto(octopi, x, y);
            }
        }

        int flashed = 0;
        for (int i = 0; i < locs->len; i++) {
            int *octo = (int*)getDictVal(octopi, locs->arr[i]);

            if (*octo >= 10) {
                *octo = 0;
                flashed++;
            }
        }

        steps++;
        if (flashed == octopi->len) {
            break;
        }
    }

    return steps;
}


int main (int argc, char *argv[]) {
    char *inputPath = defaultInput;
    if (argc > 1) {
        inputPath = argv[1];
    }

    clock_t t;
    t = clock(); 
    int p1 = part1(inputPath);
    t = clock() - t; 
    double t_p1 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 1:\nFlashes in 100 steps: %d\nRan in %f seconds\n", p1, t_p1);

    t = clock(); 
    int p2 = part2(inputPath);
    t = clock() - t;
    double t_p2 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 2:\nNumber of steps to synchronous octopi flash: %d\nRan in %f seconds\n", p2, t_p2);

    return 0;
}

File: python/11.py
import re
import sys
from pathlib import Path
from typing import Any, List, Optional, Tuple

sys.path.append(str(Path(__file__).parent.parent.parent))
from Modules.timer import Timer
def part1(data):
    """ 2021 Day 11 Part 1

    >>> part1(['5483143223', '2745854711', '5264556173', '6141336146', '6357385478', '4167524645', '2176841721', '6882881134', '4846848554', '5283751526'])
    1656
    """

    nums = [[int(x) for x in line] for line in data]
    count = 0
    for _ in range(100):
        for i in range(len(nums)):
            for j in range(len(nums[i])):
                count += updateOcto(nums, j, i)
        
        for i in range(len(nums)):
            for j in range(len(nums[i])):
                nums[i][j] = nums[i][j] % 10

    return count


def part2(data):
    """ 2021 Day 11 Part 2

    >>> part2(['5483143223', '2745854711', '5264556173', '6141336146', '6357385478', '4167524645', '2176841721', '6882881134', '4846848554', '5283751526'])
    195
    """

    nums = [[int(x) for x in line] for line in data]
    count = 0
    day = 0
    while count != len(nums) * len(nums[0]):
        day += 1
        count = 0

        for i in range(len(nums)):
            for j in range(len(nums[i])):
                nums[i][j] = nums[i][j] % 10

        for i in range(len(nums)):
            for j in range(len(nums[i])):
                count += updateOcto(nums, j, i) 

    return day


def neighborOctopi(data, x, y):
    neighbors = []
    
    for yOff in range(-1, 2):
        for xOff in range(-1, 2):
            if yOff == 0 and xOff == 0:
                continue
                
            ny = y + yOff
            nx = x + xOff
            if 0 <= ny < len(data) and 0 <= nx < len(data[0]) and [x, y] != [nx, ny]:
                neighbors.append([nx, ny])

    return neighbors


def updateOcto(data, x, y):
    if data[y][x] == 10:
        return 0

    count = 0
    data[y][x] += 1

    if data[y][x] == 10:
        count += 1
        nOctos = neighborOctopi(data, x, y)
        for n in nOctos:
            count += updateOcto(data, n[0], n[1])
    
    return count


def main(input_path: Optional[Path | str]=None, verbose: bool=False) -> Tuple[Tuple[Any, float]]:
    if not input_path:
        if not (input_path := sys.argv[1] if len(sys.argv) > 1 else None):
            year, day = re.findall(r'\d+', str(__file__))[-2:]
            input_path = Path(Path(__file__).parent.parent.parent, "Inputs", f"{year}_{day}.txt")
    
    with open(input_path, encoding='UTF-8') as f:
        data = [line.strip('\n') for line in f.readlines()]

    with Timer() as p1_time:
        p1 = part1(data)

    if verbose:
        print(f"\nPart 1:\nNumber of Octopi flashes: {p1}\nRan in {p1_time.elapsed:0.4f} seconds")

    with Timer() as p2_time:
        p2 = part2(data)

    if verbose:
        print(f"\nPart 2:\nFirst Synchronized Flash: {p2}\nRan in {p2_time.elapsed:0.4f} seconds")

    return [(p1, p1_time.elapsed), (p2, p2_time.elapsed)]


if __name__ == "__main__":
    main(verbose=True)